#1479 Add support for Builders with multiple build methods

filiphr opened the pull request:
* When there are multiple build methods look for a method named `build`, if such method exists use it
* When a builder has multiple methods throw an exception and don't use the builder during the mapping
* When a type has multiple builder creation methods throw an exception and don't use the builder

Defaulting to a method named `build` will make sure that a correct method is slected for:
* FreeBuilder - it has two methods: `build` and `buildPartial`
* Protobuf - it has three methods: `getDefaultInstanceForType`, `build` and `buildPartial`

Fixes #1479

filiphr commented:
@chris922 You also had a look at the issue, if you have time you can also review this PR. Thanks 😄

chris922 commented:
@filiphr I just had a quick look:

1. In the reference documentation you are mentioning the new exceptions. As a user of mapstruct this would be a bit confusing for me, because - as far as I got it - those exceptions are used internally and are only exposed to the user as warnings/errors during processing and nothing I can catch. So is it really helpful to mention these exception names? Maybe for a custom SPI? (haven't yet checked the builder SPI)
2. This will prevent implementing an own `@AfterMapping` that is calling the correct builder method in case one of the new exceptions would be thrown, isn't it? I still think as a user writing a custom SPI is rather complex. You may use different builder patterns in case you are using different libraries and the SPI must support everything together. So using `@AfterMapping` is for me still a good and easy solution to select the correct `build` method.

filiphr commented:
Thanks a lot @chris922.

1. I mentioned the exceptions because they are in the public SPI package and they can be thrown by a custom SPI implementations. However, you are right, maybe I can remove them from the documentation and just mention what happens when that happens (no reference to exceptions).

1. You might be right. I think that we need to move that to a separate issue. What we can do is to allow setting the methods to `null` and just passing a type. In such case we can even use an `@ObjectFactory` to create a builder instance and we probably need something like `@ObjectCreator` or whatever that can be invoked to produce a target instance. In any case I think that we need to do that in a new issue and target it for Beta2.

chris922 commented:
> MapStruct should alleviate the user from too complicated stuff like SPI's ..

I totally agree with this point. SPI should only be required for really complex stuff and not for basic stuff like deciding which `build` method should be called. 

Having a new parameter for `@Mapper`/`@MapperConfig` sounds good to me. 
Maybe this value should also be available in `@BeanMapping` to be able to decide this per mapping method and not just per mapping class?

filiphr commented:
@chris922 and @sjaakd I hear you 😄. I will try to find a better approach for this.

I agree that we need to make it easier for the users. The only issue is that the `Type` is created really early on and it's properties are required really early on, maybe even before we even touch the methods. Therefore it is a bit difficult too handle the more than one builder creation exception. Without an SPI there is no way for us to know the type for the builder and thus no way to know how to get the properties. I am open to ideas 😄 

As for the build method, I would propose that we introduce some new annotation like `@BuilderFinisher` or whatever and we can then filter in the methods that we have and if we find one we invoke that method instead.

For example instead of doing `return builder.build()` we would do `return builderFinisher(builder)`/

Btw, I am travelling until Sunday so won't be able to work on this

sjaakd commented:
> What is your suggestion about this? Is it OK to leave this for a custom SPI?

Is there a way to automatically detect what kind of builder pattern is present given the builder technology used? I think we did this recently as well, right?

filiphr commented:
>Is there a way to automatically detect what kind of builder pattern is present given the builder technology used? I think we did this recently as well, right?

By builder technology do you mean Immutables, FreeBuilder etc, or something else?

The current mechanism works in the following way:

* Find all public static methods on the type being build that have no parameters
* For each of those methods get the return type and in that type look for a public method without parameters that has a return type same as the type being build

What we can actually work with is when there is more that one builder creation method to not use builders. However, if there is a single builder creation method and multiple build methods than we would assume the user has defined a `BuilderFinisher` or `BuildFactory` and we would invoke that to finish the building. The `BuildFactory` can always be used, in the same way the `@ObjectFactory` is used.

sjaakd commented:
> By builder technology do you mean Immutables, FreeBuilder etc, or something else?

or even protobuf (believe it uses a builder pattern as well, right?



> What we can actually work with is when there is more that one builder creation method to not use builders. However, if there is a single builder creation method and multiple build methods than we would assume the user has defined a BuilderFinisher or BuildFactory and we would invoke that to finish the building. The BuildFactory can always be used, in the same way the @ObjectFactory is used.

So.. If i understand you correctly, in case of conflicts (multiple qualifying methods), you leave it to the user to point out the right "factory"  for creation, and the right "finisher" and delegate to the generated code, right? 

Could we also use the qualifier approach?  Because we normally use that to resolve conflicts somehow..?

filiphr commented:
>So.. If i understand you correctly, in case of conflicts (multiple qualifying methods), you leave it to the user to point out the right "factory" for creation, and the right "finisher" so the user could delegate to the generated code, right?


Factory for creation would be to implement an SPI as we need the builder really early, even before we work with the methods. The finisher can be done with an annotation.

>Could we also use the qualifier approach? Because we normally use that to resolve conflicts somehow..?

Not really, most of the time the builder code is generated, so there is no way for users to put the qualifiers. At least I don't see it, If you have some idea please do share it.

Example for the more than one builder creation method:

```
public Person {

    public static PersonBuilder builder() {}

    public static SecondPersonBuilder secondBuilder() {}
}
```

In the case of the builder creator for `Person` without an SPI we cannot figure out what is the type of the builder that we need, `PersonBuilder` or `SecondPersonBuilder`. If such case occurs the user will need to implement an SPI.

For the more than one build methods:

```
public class Vehicle {

    public static VehicleBuilder builder(){}

    public static class VehicleBuilder {

        public Vehicle create() {}
   
        public Vehicle otherCreate() {}
    }
}
```

In this case we can find the type of the builder (`VehicleBuilder`). However, the user will need to define something like:

```
@Mapper
public interface MyMapper {

    @BuilderFinisher
    default Vehicle build(VehicleBuilder builder) {
        return builder.otherCreate();
    }
}
```

With `@BuilderFinisher` the users would be able to define their own way of invoking the builder.

How does this sound?

sjaakd commented:
Ill check it out this evening.. 



> Op 3 jun. 2018 om 15:18 heeft Filip Hrisafov <notifications@github.com> het volgende geschreven:
> 
> So.. If i understand you correctly, in case of conflicts (multiple qualifying methods), you leave it to the user to point out the right "factory" for creation, and the right "finisher" so the user could delegate to the generated code, right?
> 
> Factory for creation would be to implement an SPI as we need the builder really early, even before we work with the methods. The finisher can be done with an annotation.
> 
> Could we also use the qualifier approach? Because we normally use that to resolve conflicts somehow..?
> 
> Not really, most of the time the builder code is generated, so there is no way for users to put the qualifiers. At least I don't see it, If you have some idea please do share it.
> 
> Example for the more than one builder creation method:
> 
> public Person {
> 
>     public static PersonBuilder builder() {}
> 
>     public static SecondPersonBuilder secondBuilder() {}
> }
> In the case of the builder creator for Person without an SPI we cannot figure out what is the type of the builder that we need, PersonBuilder or SecondPersonBuilder. If such case occurs the user will need to implement an SPI.
> 
> For the more than one build methods:
> 
> public class Vehicle {
> 
>     public static VehicleBuilder builder(){}
> 
>     public static class VehicleBuilder {
> 
>         public Vehicle create() {}
>    
>         public Vehicle otherCreate() {}
>     }
> }
> In this case we can find the type of the builder (VehicleBuilder). However, the user will need to define something like:
> 
> @Mapper
> public interface MyMapper {
> 
>     @BuilderFinisher
>     default Vehicle build(VehicleBuilder builder) {
>         return builder.otherCreate();
>     }
> }
> With @BuilderFinisher the users would be able to define their own way of invoking the builder.
> 
> How does this sound?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

sjaakd commented:
@filiphr .. I understand your approach.

However it would still take a considerable  amount of work to do this for each method. If you only get 3 or 4 properties to map in a bean, you now need to implement possibly a creater and a finisher method. Do you know if the builders predominantly use a consistent naming convention that we could use? E.g. if its always `createXYZ` this could be considered a pattern. You could also do this on the `@MappingConfig`, `@Mapping` and `@BeanMapping` as two properties. Like `builderCreatetor = "^createSecond.*$"`, `builderFinisher = "^buildSecond.*$"`.

filiphr commented:
@sjaakd, @chris922 I've changed the approach a bit. I took ideas from you, @sjaakd . There is now `@BuilderMapping` in which you can set the pattern for the finisher method.

I still need to update the Javadoc and the reference documentation, but wanted to get your opinion first. I went with a standalone annotation as that opens the path to meta annotations for 2.0 (perhaps). 

WDYT?

sjaakd commented on core-common/src/main/java/org/mapstruct/BuilderMapping.java:
```diff
@@ -0,0 +1,38 @@
+})
+public @interface BuilderMapping {
+
+    String builderFinisher() default "build";
```
I thought the problem was 2 fold: `builderFinisher` and `builderCreator`.. where's the latter one?

filiphr replied:
The `builderCreator` one is a more complex one. We need the type of the builder at the time of the creation of the `Type` itself. As we need the properties really early on, when we are constructing the `Target` and `Source` references. That's why that is not hear, and I am having hard time pulling that here.

sjaakd replied:
ok.. but then leave it open as a new issue / improvement.. From a user perspective, its strange, agreed?

filiphr replied:
I'll create an issue for that. yes it might be a bit strange from a user perspective.

filiphr replied:
Issue created https://github.com/mapstruct/mapstruct/issues/1547

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/BuilderType.java:
```diff
@@ -107,11 +107,6 @@ public static BuilderType create(BuilderInfo builderInfo, Type typeToBuild, Type
         if ( builderInfo == null ) {
             return null;
         }
-        ExecutableElement buildMethod = builderInfo.getBuildMethod();
```
is this check obsolete?

filiphr replied:
Probably this check belongs to the `BuilderFinisherMethodResolver` now. As we can check the method return type there. Although this can't happen with the default builder provider as we have this check there.

sjaakd replied:
ok

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/BuilderFinisherMethodResolver.java:
```diff
@@ -0,0 +1,109 @@
+            }
+            for ( ExecutableElement buildMethod : buildMethods ) {
+                String methodName = buildMethod.getSimpleName().toString();
+                if ( methodName.matches( buildMethodPattern ) ) {
```
Yes we could. However, currently we are making sure that this can't happen in the `DefaultBuilderProvider`. We can even make that as part of the contract of the `BuilderInfo`.

sjaakd replied:
ok

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/BuilderFinisherMethodResolver.java:
```diff
@@ -0,0 +1,109 @@
+        }
+        else {
+            BuilderMappingPrism builderMapping = builderMappingPrism( method, ctx );
+            String buildMethodPattern = "build";
```
Makes sense

filiphr commented on core-common/src/main/java/org/mapstruct/BuilderMapping.java:
```diff
@@ -0,0 +1,38 @@
+    ElementType.TYPE,
+    ElementType.METHOD
+})
+public @interface BuilderMapping {
```
You are right. It might not make sense to `MapMapping`. However, it will make sense during the computation of the intermediate methods for the keys / values mappings. In my opinion we need to change the `MapperConfig` and allow use of meta annotations in 2.0. The mapper config might even not be needed.

filiphr replied:
Yes we do that in 2.0. I will apply the changes here.

filiphr commented:
@sjaakd it might actually make sense to apply it on `MapMapping` or `IterableMapping`. The reason is that we are constructing intermediary methods in there, so if they need builders we can use that from there (although I think it is not passed appropriately now).

2. Yes my idea was to go with meta annotations for the future. In any case I can change this and apply the `BuilderMapping` as a field in the `Mapper`, `MapperConfig` and `BeanMapping` in order to be consistent with the current paradigm.

How should I proceed with this one?

sjaakd commented:
> How should I proceed with this one?

I would prefer to have a consists meta annotation interface in 2.0 and (a consistent one now) :smile:

chris922 commented on processor/src/main/java/org/mapstruct/ap/internal/model/BuilderFinisherMethodResolver.java:
```diff
@@ -0,0 +1,109 @@
+            //If we reach this method this should never happen
+            return null;
+        }
+        else if ( buildMethods.size() == 1 ) {
```
If a `@BuilderMapping` was defined with an explicit name for the `builderFinisher`, but only a different method was found, the wrong method will be used. The user would not notice this.

In case he entered a `builderFinisher` I would expect a warning or even better an error in case the only method found has a different name. The issue is most likely how we should know that he explicitly entered  a name, because you defined `build` as a default name in the annotation. So we will always have a `builderFinisher` as soon as the `@BuilderMapping` annotation _(or whatever it will be at the end)_ will be used.

The only way I see to solve this issue is to use an empty string as a default annotation attribute value and in this case use `build` as a constant value here in the code (the `DEFAULT_BUILD_METHOD_NAME`).
And switch the `if-else` ordering.. if a `builderFinisher` was defined try to use a method with exactly this name, if not possible -> error. If nothing was defined fallback to the way you are doing it right now (`size == 1`, then `methodName equals build`).

chris922 commented on processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:
```diff
@@ -100,6 +100,10 @@
     GENERAL_NO_SUITABLE_CONSTRUCTOR( "%s does not have an accessible parameterless constructor." ),
 
+    BUILDER_MORE_THAN_ONE_BUILDER_CREATION_METHOD( "More than one builder creation method for \"%s\". Found methods: \"%s\". Builder will not be used. Consider implementing a custom BuilderProvider SPI.", Diagnostic.Kind.WARNING ),
+    BUILDER_MORE_THAN_ONE_BUILD_METHOD_DEFAULT_PATTERN( "More than one build method in \"%s\" for \"%s\". Found methods: \"%s\". Using default \"build\" method pattern. Consider to add @BuilderMapping in order to select the correct build method.", Diagnostic.Kind.ERROR ),
```
This _"Using default .. method pattern"_ sounds confusing for me. Just after reading this message I thought this is an INFO that he found more than one possible build method, but uses at the end the one with the name `build`. Maybe something like:

`More than one build method in \"%s\" for \"%s\". Found methods: \"%s\", but no method is named \"build\". Consider to add @BuilderMapping in order to select the correct build method.`

chris922 commented on processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:
```diff
@@ -100,6 +100,10 @@
 
+    BUILDER_MORE_THAN_ONE_BUILDER_CREATION_METHOD( "More than one builder creation method for \"%s\". Found methods: \"%s\". Builder will not be used. Consider implementing a custom BuilderProvider SPI.", Diagnostic.Kind.WARNING ),
+    BUILDER_MORE_THAN_ONE_BUILD_METHOD_DEFAULT_PATTERN( "More than one build method in \"%s\" for \"%s\". Found methods: \"%s\". Using default \"build\" method pattern. Consider to add @BuilderMapping in order to select the correct build method.", Diagnostic.Kind.ERROR ),
+    BUILDER_MORE_THAN_ONE_BUILD_METHOD_DEFINED_PATTERN( "More than one build method in \"%s\" for \"%s\". Found methods: \"%s\". Using defined \"%s\" method pattern.", Diagnostic.Kind.ERROR ),
```
Same as above, maybe:

`More than one build method in \"%s\" for \"%s\". Found methods: \"%s\", but no method is named \"%s\" as defined in @BuilderMapping#builderFinisher.`

filiphr commented:
I've finally found the time to go over the changes in this PR 😄.

Thank you @chris922 and @sjaakd for the reviews.

I've applied both of your comments.

* We now have the `@BuilderMapping` as part of the `@BeanMapping`, `@Mapper` and/or `@MapperConfig`
* I've adapted the error messages a bit, they should be more descriptive now.

If you have time let me know what you think.

sjaakd commented:
Hey Filip.. Looks more consistent. 

I'm still a bit struggling whether we want to have annotations within annotations to solve this problem. I know you are considering some kind of meta annotation mechanism for 2.0. 

However, It would be more usable (and step-up to 2.0) if you could define an annotation as a user (a bit like we do with qualifiers), grouping specific `@BuilderMapping ` properties for reuse and mark that annotation with `@BuilderMapping`

So:
```java
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
@BuilderMapping(builderFinisher = "create")
public @interface BuilderMappingForFrameworkX { }
```

```java
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
@BuilderMapping(builderFinisher = "wrongCreate")
public @interface BuilderMappingForFrameworkY { }
```

You should then also be able to use the `@BuilderMappingForFrameworkX` and `@BuilderMappingForFrameworkY ` directly :

```java
@Mapper(builder = @BuilderMappingForFrameworkX)
public interface BuilderMappingDefinedMapper {

    BuilderMappingDefinedMapper INSTANCE = Mappers.getMapper( BuilderMappingDefinedMapper.class );

    Process map(Source source);

    @BeanMapping(builder = @BuilderMappingForFrameworkY)
    Process wrongMap(Source source);
}
```

If this is possible, we could:
1. Extend this mechanism to other places (e.g. reusable pieces of mappings instead of the `@InheritConfiguration` / `@InheritInverseConfiguration`
2. Define our own extra's... e.g. `@BuilderMappingForProtoBuf`, `@BuilderMappingForLombok`

Note: Having only one property in `@BuilderMapping` makes it a bit hard to see where you are heading for.  

Shall we discuss this in our Tuesday meeting?

gunnarmorling commented on core-common/src/main/java/org/mapstruct/BuilderMapping.java:
```diff
@@ -0,0 +1,42 @@
+import java.lang.annotation.Target;
+
+/**
+ * Annotation that can be used to provide a custom finisher for the builders in case of multiple build methods.
```
Let's not repeat that it's an annotation, that's clear from its definition:

> Configuration of builders, e.g. the name of the final build method.

gunnarmorling commented on core-common/src/main/java/org/mapstruct/BuilderMapping.java:
```diff
@@ -0,0 +1,42 @@
+ */
+@Retention(RetentionPolicy.CLASS)
+@Target({})
+public @interface BuilderMapping {
```
Should it be marked as `@Incubating` or are we sure that it won't be altered incompatibly down the road?

filiphr commented:
I've squashed all commits, did the rename, update the commit message to:

> * Add new @builder annotation for defining a build method
>* When there are multiple build methods look for a method named `build` and if found use it
>* If @builder is defined than look for a build method with the defined method
>* When a type has multiple builder creation methods throw an exception and don't use the builder
>
>Defaulting to a method named `build` will make sure that a correct method is selected for:
>* FreeBuilder - it has two methods: `build` and `buildPartial`
>* Protobuf - it has three methods: `getDefaultInstanceForType`, `build` and `buildPartial`

I also updated the documentation. Once the build is done, I'll do the merge

chris922 commented on core-common/src/main/java/org/mapstruct/Builder.java:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * The name of the build method that needs to be invoked on the builder to create the type being build
+     *
+     * @return the method that needs to tbe invoked on the builder
```
`to be` :)

chris922 commented on core-common/src/main/java/org/mapstruct/Mapper.java:
```diff
@@ -200,4 +200,22 @@
+     * NOTE: In case no builder is defined here, in {@link BeanMapping} or {@link MapperConfig} and there is a single
+     * build method, then that method would be used.
+     * <p>
+     * If the builder is defined and there is a single method that does not match the name of the finisher than
```
`the name of the build method` as we don't use `finisher` anymore?

chris922 commented on core-common/src/main/java/org/mapstruct/Mapper.java:
```diff
@@ -200,4 +200,22 @@
+     * build method, then that method would be used.
+     * <p>
+     * If the builder is defined and there is a single method that does not match the name of the finisher than
+     * a compile error will occurs
```
`occur.`

chris922 commented on core-common/src/main/java/org/mapstruct/MapperConfig.java:
```diff
@@ -186,4 +186,24 @@ MappingInheritanceStrategy mappingInheritanceStrategy()
+     * @return the builder information
+     *
+     * @since 1.3
+     */
```
same here

chris922 commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -639,6 +639,12 @@ The default implementation of the `BuilderProvider` assumes the following:
 So for example `Person` has a public static method that returns `PersonBuilder`.
 * The builder type has a parameterless public method (build method) that returns the type being build
 In our example `PersonBuilder` has a method returning `Person`.
+* In case there are multiple build methods, MapStruct will look for a method called `build` if such methods exists
```
`method`

filiphr commented:
Thanks for the final checks @chris922. I'll commit those directly on master 😄
